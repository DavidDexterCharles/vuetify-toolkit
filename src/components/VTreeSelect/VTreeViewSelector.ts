import Vue, { VueConstructor } from 'vue'
import { VTreeview } from '../../vuetify-import'

const VTreeviewComponent = VTreeview as VueConstructor<Vue>

export default VTreeviewComponent.extend({
  name: 'v-tree-view-selector',
  props: {
    selectedItems: {
      type: Array,
      default: []
    }
  },
  data: () => ({
    selectedCache: new Set()
  }),
  watch: {
    selectedItems () {
      if (this.selectedItems.length < 1) {
        this.clearSelection()
      }
    }
  },
  methods: {
    clearSelection () {
      this.selectedCache.clear()
      Object.keys(this.$data.nodes).forEach(key => {
        this.$data.nodes[key].isSelected = false
        this.$data.nodes[key].isIndeterminate = false
      })
    },
    selectAll () {
      this.selectedCache.clear()
      Object.keys(this.$data.nodes).forEach(key => {
        this.$data.nodes[key].isSelected = true
        this.$data.nodes[key].isIndeterminate = false
      })
    }
  }
})
