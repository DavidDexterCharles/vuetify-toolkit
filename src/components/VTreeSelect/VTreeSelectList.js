import { mixins, VTreeviewNodeProps, VListItem, VListItemContent, VListItemTitle, Themeable, Colorable } from '../../vuetify-import'

import VTreeViewSelector from './VTreeViewSelector'

export default mixins(
  Themeable, Colorable
  /* @vue/component */
).extend({
  name: 'v-tree-select-list',
  props: {
    selectedItems: {
      type: Array,
      default: () => ([])
    },
    noDataText: String,
    dense: Boolean,
    multiple: Boolean,
    items: {
      type: Array,
      default: () => ([])
    },
    openAll: Boolean,
    returnObject: {
      type: Boolean,
      default: false // TODO: Should be true in next major
    },
    value: {
      type: Array,
      default: () => ([])
    },
    search: String,
    filter: {
      type: Function,
      default: undefined
    },
    useDefaultToolbarCommand: {
      type: Boolean,
      default: false
    },
    // custom commands
    toolbarCommands: {
      type: Array,
      default: () => []
    },
    ...VTreeviewNodeProps
  },
  computed: {
    staticNoDataTile () {
      const tile = {
        on: {
          mousedown: e => e.preventDefault() // Prevent onBlur from being called
        }
      }
      return this.$createElement(VListItem, tile, [
        this.genTileNoDataContent()
      ])
    }
  },
  methods: {
    genTileNoDataContent () {
      const innerHTML = this.noDataText
      return this.$createElement(VListItemContent,
        [this.$createElement(VListItemTitle, {
          domProps: { innerHTML }
        })]
      )
    }
  },
  render () {
    const children = []
    if (!this.items || !Array.isArray(this.items) || this.items.length < 1) {
      children.push(this.$slots['no-data'] || this.staticNoDataTile)
    }
    this.$slots['prepend-item'] && children.unshift(this.$slots['prepend-item'])
    const childrenAppend = []
    this.$slots['append-item'] && childrenAppend.push(this.$slots['append-item'])

    return this.$createElement('div', {
      staticClass: 'v-select-list v-card',
      'class': this.themeClasses
    }, [
      children,
      this.$createElement(VTreeViewSelector, {
        props: {
          dense: this.dense,
          items: this.items,
          selectable: true,
          returnObject: true,
          selectOnly: true,
          selectedItems: this.selectedItems,
          openAll: this.openAll,
          itemText: this.itemText
        },
        on: {
          input: e => {
            this.$emit('select', e)
          }
        }
      }), childrenAppend
    ])
  }
})
