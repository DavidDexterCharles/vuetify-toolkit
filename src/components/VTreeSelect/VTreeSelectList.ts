import { VNode, PropType } from 'vue'
import { mixins, VTreeviewNodeProps, VListItem, VListItemContent, VListItemTitle, Themeable, Colorable } from '../../vuetify-import'

import VTreeViewSelector from './VTreeViewSelector'

const mixinsComponent = mixins as any

export default mixinsComponent(
  Themeable, Colorable
  /* @vue/component */
).extend({
  name: 'v-tree-select-list',
  props: {
    selectedItems: {
      type: Array,
      default: () => ([])
    },
    noDataText: String,
    dense: Boolean,
    multiple: Boolean,
    items: {
      type: Array,
      default: () => ([])
    },
    openAll: Boolean,
    returnObject: {
      type: Boolean,
      default: false // TODO: Should be true in next major
    },
    value: {
      type: Array,
      default: () => ([])
    },
    search: String,
    filter: {
      type: Function,
      default: undefined
    },
    useDefaultToolbarCommand: {
      type: Boolean,
      default: false
    },
    // custom commands
    toolbarCommands: {
      type: Array,
      default: () => []
    },
    ...VTreeviewNodeProps,
    selectionType: {
      type: String as PropType<'leaf' | 'independent'>,
      default: 'leaf',
      validator: (v: string) => ['leaf', 'independent'].includes(v)
    }
  },
  computed: {
    staticNoDataTile (): VNode {
      const tile = {
        on: {
          mousedown: (e: any) => e.preventDefault() // Prevent onBlur from being called
        }
      }
      return (this as any).$createElement(VListItem, tile, [
        (this as any).genTileNoDataContent()
      ])
    }
  },
  methods: {
    genTileNoDataContent (): VNode {
      const innerHTML = (this as any).noDataText
      return (this as any).$createElement(VListItemContent,
        [(this as any).$createElement(VListItemTitle, {
          domProps: { innerHTML }
        })]
      )
    }
  },
  render (): VNode {
    const children = []
    if (!this.items || !Array.isArray(this.items) || this.items.length < 1) {
      children.push(this.$slots['no-data'] || this.staticNoDataTile)
    }
    this.$slots['prepend-item'] && children.unshift(this.$slots['prepend-item'])
    const childrenAppend = []
    this.$slots['append-item'] && childrenAppend.push(this.$slots['append-item'])

    return this.$createElement('div', {
      staticClass: 'v-select-list v-card',
      class: this.themeClasses
    }, [
      children,
      this.$createElement(VTreeViewSelector, {
        props: {
          dense: this.dense,
          items: this.items,
          itemDisabled: this.itemDisabled,
          itemText: this.itemText,
          itemKey: this.itemKey,
          itemChildren: this.itemChildren,
          loadChildren: this.loadChildren,
          selectable: true,
          returnObject: true,
          selectOnly: true,
          selectedItems: this.selectedItems,
          openAll: this.openAll,
          shaped: this.shaped,
          rounded: this.rounded,
          selectionType: this.selectionType
        },
        scopedSlots: this.$scopedSlots,
        on: {
          input: (e: any[]) => {
            this.$emit('select', e)
          }
        }
      }), childrenAppend
    ])
  }
})
