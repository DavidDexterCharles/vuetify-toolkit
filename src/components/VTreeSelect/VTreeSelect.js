import { VAutocomplete, VSelect, VTreeviewNodeProps, consoleError } from '../../vuetify-import'
import VTreeSelectList from './VTreeSelectList'
import DefaultMenuProps from '../../utils/MenuProps'
import treeviewScopedSlots from '../../utils/TreeviewScopedSlots'

export default VAutocomplete.extend({
  name: 'v-tree-select',
  props: {
    ...VSelect.options.props,
    ...VAutocomplete.options.props,
    ...VTreeviewNodeProps,
    autocomplete: {
      type: Boolean,
      default: false
    },
    menuProps: {
      type: [String, Array, Object],
      default: () => DefaultMenuProps
    },
    openAll: Boolean
  },
  data: () => ({
    selectedItems: []
  }),
  computed: {
    classes () {
      if (this.autocomplete) {
        return Object.assign({}, VSelect.options.computed.classes.call(this), {
          'v-autocomplete': true,
          'v-autocomplete--is-selecting-index': this.selectedIndex > -1
        })
      } else {
        return Object.assign({}, VSelect.options.computed.classes.call(this), {})
      }
    },
    internalSearch: {
      get () {
        const result = this.autocomplete ? VAutocomplete.options.computed.internalSearch.get.call(this)
          : ''
        return result
      },
      set (val) {
        if (this.autocomplete) {
          VAutocomplete.options.computed.internalSearch.set.call(this, val)
        }
      }
    },
    listData () {
      const data = VSelect.options.computed.listData.call(this)
      Object.assign(data.props, { ...VTreeviewNodeProps })
      /* to remove console warns and type conflicts */
      Object.assign(data.props, {
        activatable: this.activatable,
        activeClass: this.activeClass,
        color: this.color,
        dark: this.dark,
        selectable: true,
        selectedColor: this.selectedColor,
        indeterminateIcon: this.indeterminateIcon,
        onIcon: this.onIcon,
        offIcon: this.offIcon,
        expandIcon: this.expandIcon,
        loadingIcon: this.loadingIcon,
        loadChildren: this.loadChildren,
        itemKey: this.itemKey,
        itemText: this.itemText,
        itemChildren: this.itemChildren,
        itemDisabled: this.itemDisabled,
        transition: this.transition,
        selectedItems: this.selectedItems,
        shaped: this.shaped,
        rounded: this.rounded,
        openAll: this.openAll,
        openOnClick: this.openOnClick,
        useDefaultToolbarCommand: this.useDefaultToolbarCommand,
        toolbarCommands: this.toolbarCommands
      })
      Object.assign(data.on, {
        select: e => {
          this.selectItems(e)
        }
      })
      Object.assign(data.scopedSlots, treeviewScopedSlots(this.$scopedSlots))
      return data
    },
    staticList () {
      if (this.$slots['no-data'] || this.$slots['prepend-item'] || this.$slots['append-item']) {
        consoleError('assert: staticList should not be called if slots are used')
      }
      return this.$createElement(VTreeSelectList, this.listData)
    }
  },
  methods: {
    register () {},
    genInput () {
      return this.autocomplete ? VAutocomplete.options.methods.genInput.call(this)
        : VSelect.options.methods.genInput.call(this)
    },
    genListWithSlot () {
      const slots = ['prepend-item', 'no-data', 'append-item']
        .filter(slotName => this.$slots[slotName])
        .map(slotName => this.$createElement('template', {
          slot: slotName
        }, this.$slots[slotName]))
      // Requires destructuring due to Vue
      // modifying the `on` property when passed
      // as a referenced object
      return this.$createElement(VTreeSelectList, {
        ...this.listData
      }, slots)
    },
    genSelections () {
      let length = this.selectedItems.length
      const children = new Array(length)

      let genSelection
      if (this.$scopedSlots.selection) {
        genSelection = this.genSlotSelection
      } else if (this.hasChips) {
        genSelection = this.genChipSelection
      } else {
        genSelection = this.genCommaSelection
      }

      while (length--) {
        children[length] = genSelection(
          this.selectedItems[length],
          length,
          length === children.length - 1
        )
      }

      return this.$createElement('div', {
        staticClass: 'v-select__selections'
      }, children)
    },
    selectItems (items) {
      this.selectedItems = items
      if (!this.multiple) {
        this.isMenuActive = false
      }
    },
    clearableCallback () {
      this.internalValue = null
      this.$refs.input.internalValue = ''
      this.$refs.input.value = ''
      this.selectedItems = []
      this.$nextTick(() => this.$refs.input.focus())
    }
  }
})
