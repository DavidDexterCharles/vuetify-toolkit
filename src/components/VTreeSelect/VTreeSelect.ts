import Vue, { VNode, VueConstructor, VNodeData } from 'vue'
import { VAutocomplete, VSelect, VTreeviewNodeProps, consoleError } from '../../vuetify-import'
import VTreeSelectList from './VTreeSelectList'
import DefaultMenuProps from '../../utils/MenuProps'
import treeviewScopedSlots from '../../utils/TreeviewScopedSlots'

const VAutocompleteComponent = VAutocomplete as VueConstructor<Vue>

export default VAutocompleteComponent.extend({
  name: 'v-tree-select',
  props: {
    ...(VSelect as any).options.props,
    ...(VAutocomplete as any).options.props,
    ...VTreeviewNodeProps,
    autocomplete: {
      type: Boolean,
      default: false
    },
    menuProps: {
      type: [String, Array, Object],
      default: () => DefaultMenuProps
    },
    openAll: Boolean
  },
  data: () => ({
    selectedItems: [] as any[]
  }),
  computed: {
    classes (): Object {
      if (this.$props.autocomplete) {
        return Object.assign({}, (VSelect as any).options.computed.classes.call(this), {
          'v-autocomplete': true,
          'v-autocomplete--is-selecting-index': this.$data.selectedIndex > -1
        })
      } else {
        return Object.assign({}, (VSelect as any).options.computed.classes.call(this), {})
      }
    },
    internalSearch: {
      get (): string {
        const result = this.$props.autocomplete ? (VAutocomplete as any).options.computed.internalSearch.get.call(this)
          : ''
        return result
      },
      set (val: string) {
        if (this.$props.autocomplete) {
          (VAutocomplete as any).options.computed.internalSearch.set.call(this, val)
        }
      }
    },
    listData (): Object {
      const data = (VSelect as any).options.computed.listData.call(this)
      Object.assign(data.props, { ...VTreeviewNodeProps })
      /* to remove console warns and type conflicts */
      Object.assign(data.props, {
        activatable: this.$props.activatable,
        activeClass: this.$props.activeClass,
        color: this.$props.color,
        chips: this.$props.chips,
        dark: this.$props.dark,
        selectable: true,
        selectedColor: this.$props.selectedColor,
        indeterminateIcon: this.$props.indeterminateIcon,
        onIcon: this.$props.onIcon,
        offIcon: this.$props.offIcon,
        expandIcon: this.$props.expandIcon,
        loadingIcon: this.$props.loadingIcon,
        loadChildren: this.$props.loadChildren,
        itemKey: this.$props.itemKey,
        itemText: this.$props.itemText,
        itemChildren: this.$props.itemChildren,
        itemDisabled: this.$props.itemDisabled,
        transition: this.$props.transition,
        selectedItems: this.selectedItems,
        shaped: this.$props.shaped,
        rounded: this.$props.rounded,
        openAll: this.$props.openAll,
        openOnClick: this.$props.openOnClick,
        useDefaultToolbarCommand: this.$props.useDefaultToolbarCommand,
        toolbarCommands: this.$props.toolbarCommands
      })
      Object.assign(data.on, {
        select: (e: any[]) => {
          this.selectItems(e)
        }
      })
      Object.assign(data.scopedSlots, treeviewScopedSlots(this.$scopedSlots))
      return data
    },
    staticList (): VNode {
      if (this.$slots['no-data'] || this.$slots['prepend-item'] || this.$slots['append-item']) {
        consoleError('assert: staticList should not be called if slots are used')
      }
      return this.$createElement(VTreeSelectList, this.listData)
    }
  },
  methods: {
    register () {},
    genInput (): VNode {
      return this.$props.autocomplete ? (VAutocomplete as any).options.methods.genInput.call(this)
        : (VSelect as any).options.methods.genInput.call(this)
    },
    genListWithSlot (): VNode {
      const slots = ['prepend-item', 'no-data', 'append-item']
        .filter(slotName => this.$slots[slotName])
        .map(slotName => this.$createElement('template', {
          slot: slotName
        }, this.$slots[slotName]))
      // Requires destructuring due to Vue
      // modifying the `on` property when passed
      // as a referenced object
      return this.$createElement(VTreeSelectList, ({ ...this.listData }) as VNodeData, slots)
    },
    genSelections (): VNode {
      let length = this.selectedItems.length
      const children = new Array(length)

      let genSelection
      if (this.$scopedSlots.selection) {
        genSelection = (this as any).genSlotSelection
      } else if (this.$props.hasChips) {
        genSelection = (this as any).genChipSelection
      } else {
        genSelection = (this as any).genCommaSelection
      }

      while (length--) {
        children[length] = genSelection(
          this.selectedItems[length],
          length,
          length === children.length - 1
        )
      }

      return this.$createElement('div', {
        staticClass: 'v-select__selections'
      }, children)
    },
    selectItems (items: any[]) {
      this.selectedItems = items
      if (!this.$props.multiple) {
        this.$data.isMenuActive = false
      }
    },
    clearableCallback () {
      this.$data.internalValue = null;
      // this.$refs.input.internalValue = '';
      (this.$refs.input as HTMLInputElement).value = ''
      this.selectedItems = []
      this.$nextTick(() => (this.$refs.input as HTMLInputElement).focus())
    }
  }
})
