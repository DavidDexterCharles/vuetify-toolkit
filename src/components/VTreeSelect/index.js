import VTreeSelect from './VTreeSelect'

export default VTreeSelect
