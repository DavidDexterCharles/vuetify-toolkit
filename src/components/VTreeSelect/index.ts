import VTreeSelect from './VTreeSelect'

export { VTreeSelect }
export default VTreeSelect
