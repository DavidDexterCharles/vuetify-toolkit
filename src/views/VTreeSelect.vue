<template>
  <div>
      <h1>VTreeSelect</h1>
      <span></span>
      <p>VTreeSelect is the component that provide tree-like data selection. This combines original Vuetify components VAutocomplete and VTreeview in one.
        You can use any nested data like item source.
      </p>
      <h2>API</h2>
      <v-card>
        <v-api :PROPS="componentProps" :SLOTS="componentSlots" :EVENTS="componentEvents">
        </v-api>
      </v-card>
      <span/>
      <h2>Sandbox</h2>
      <v-card :dark="dark">
        <v-card-text>
          <v-row justify="space-around">
          <v-switch v-model="multiple" class="ma-2" label="multiple"></v-switch>
          <v-switch v-model="chips" class="ma-2"    label="chips"></v-switch>
          <v-switch v-model="dense" class="ma-2"    label="dense"></v-switch>
          <v-switch v-model="clearable" class="ma-2"    label="clearable"></v-switch>
          <v-switch v-model="dark" class="ma-2"    label="dark"></v-switch>
          </v-row>
          <v-tree-select :chips = "chips" :dense = "dense" :multiple = "multiple" :items="items" :clearable = "clearable" :dark = "dark"></v-tree-select>
        </v-card-text>
      </v-card>
      <span/>
      <h2>Examples</h2>
      <h3>v-tree-select-select example (chips selections)</h3>
      <v-tree-select chips :items="items" clearable></v-tree-select>
      <h3>v-tree-select-select example (no data)</h3>
      <v-tree-select chips clearable></v-tree-select>
      <h3>v-tree-select-select example (multiple selection)</h3>
      <v-tree-select chips :items="items" clearable multiple></v-tree-select>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import VTreeSelect from '../components/VTreeSelect'
import VApi from './API.vue'
import { staticitems, dataGridHeaders, desserts } from '../example-data'

export default Vue.extend({
  components: {
    VTreeSelect,
    VApi
  },
  data: () => ({
    items: staticitems,
    dataGridHeaders: dataGridHeaders,
    desserts: desserts,
    chips: false,
    multiple: false,
    dense: false,
    clearable: false,
    dark: false,
    //
    componentProps: [
      { name: 'autocomplete',
        type: 'boolean',
        default: 'true',
        describtion: ' enable or disable text typing for autocomplete behavior'
      },
      { name: 'dense',
        type: 'boolean',
        default: 'true',
        describtion: ' enable or disable text typing for autocomplete behavior'
      }
    ],
    //
    componentSlots: [
      { name: 'default',
        describtion: ' default component slot '
      }
    ],
    //
    componentEvents: [
      { name: 'default',
        describtion: ' default component slot ',
        value: ''
      }
    ]
  })
})
</script>
