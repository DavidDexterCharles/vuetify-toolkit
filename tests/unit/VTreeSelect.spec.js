// Libraries
import Vue from 'vue'
import Vuetify from 'vuetify'

// Utilities
import {
  mount,
  createLocalVue
} from '@vue/test-utils'

import VTreeSelect from '@/components/VTreeSelect/VTreeSelect'

// console.log(vuetify)

const localVue = createLocalVue()

const singleRootTwoChildren = [
  { id: 0, name: 'Root', children: [{ id: 1, name: 'Child' }, { id: 2, name: 'Child 2' }] }
]

const threeLevels = [
  { id: 0, name: 'Root', children: [{ id: 1, name: 'Child', children: [{ id: 2, name: 'Grandchild' }] }, { id: 3, name: 'Child' }] }
]

describe('VTreeSelect.js', () => {
  // let vuetify

  // beforeEach(() => {
  //   vuetify = new Vuetify()
  // })
  // console.log(vuetify)

  let vuetify = new Vuetify({
      $vuetify: {
        lang: {
          t: (val) => val
        }
      }
  })

  console.log(vuetify)

  it('renders all items', () => {
    const wrapper = mount(VTreeSelect, {
      vuetify,
      propsData: {
        items: singleRootTwoChildren
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
