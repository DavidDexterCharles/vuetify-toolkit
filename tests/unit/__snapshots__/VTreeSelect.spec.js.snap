// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTreeSelect.js renders 1`] = `
<div class="v-input theme--light v-text-field v-select">
  <div class="v-input__control">
    <div role="combobox" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-1" class="v-input__slot">
      <div class="v-select__slot">
        <div class="v-select__selections"><input id="input-1" readonly="readonly" type="text" aria-readonly="true"></div>
        <div class="v-input__append-inner">
          <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
        </div><input type="hidden">
      </div>
      <div class="v-menu">
        <div class="v-menu__content theme--light v-autocomplete__content" style="max-height: auto; min-width: 0px; max-width: auto; top: 12px; left: 0px; transform-origin: top left; z-index: 0; display: none;"></div>
      </div>
    </div>
    <div class="v-text-field__details">
      <div class="v-messages theme--light"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
    </div>
  </div>
</div>
`;
